app.py

```diff
@@ -1007,6 +1007,170 @@ def set_routing_time_maximum(self):
+            dest:
+                Contains the destination PID
+        OUTPUTS-----
+            reduction_cost:
+                returns the additional reduction cost that results in the path being taken
+            matrix:
+                Returns a reduced matrix with the specificed row and columns reduced
+        """
+        row_cost, col_cost = INFINITY
+        reduction_cost = 0;
+        # Row Reduction
+        for k,v in matrix.items():
+            if (source == k[0]):
+        
+                for direc in v:
+                    direc_cost = INFINITY if (v.get(direc).get('cost') is None) else v.get(direc).get('cost')
+                    row_cost = min(row_cost, direc_cost)
+    
+        for k,v in matrix.items():
+            if (source == k[0]):
+                
+                for direc in v:
+                    v[direc]['cost'] = None if (v.get(direc).get('cost') is None) else (v.get(direc).get('cost') - row_cost)
+        reduction_cost += row_cost
+
+        # Column Reduction
+        for k,v in matrix.items():
+            if (dest == k[1]):
+
+                for direc in v:
+                    direc_cost = INFINITY if (v.get(direc).get('cost') is None) else v.get(direc).get('cost')
+                    col_cost = min(col_cost, direc_cost)
+        
+        for k,v in matrix.items():
+            if (dest == k[1]):
+                for direc in v:
+         
+                    v[direc]['cost'] = None if (v.get(direc).get('cost') is None) else (v.get(direc).get('cost') - col_cost)
+        reduction_cost += col_cost
+
+        return reduction_cost, matrix
+
+"""
+    def dfs(self, item_list):
+        
+        Performs DFS on a list to produce all permutations, use for brute force to find all paths in the tree.
+        
+        if (len(item_list) == 0):
+            return []
+        if (len(item_list) == 1):
+            return []
+
+        result = []
+        for i in range(len(item_list)):
+            node = item_list[i]
+            
+            # Pops the other access point of the chosen item so there are no repeat vists for the same item
+            for k,v in item_list.items():
+                if (node[0] == k[0] and node[2] != k[2]):
+                    item_list.pop()
+
+            remaining_item_list = item_list[:i] + item_list[i+1:]
+
+            for path in defs(remaining_item_list):
+                result.append([node] + path)
+
+        return result
+"""
+
+    def branch_n_bound(self, graph):
+        """
+        Applies the branch and bound algorithm to generate a path
+        """
+        path = []
+        # upper_bound = 
+        # path_complete = len( graph.keys() )   # number of products in the order
+        rand_start = random.choice( graph.keys() )
+        path.append( (rand_start[0], rand_start[2]) ) # (source, source access direction)
+
+        # Creates the reduced matrix
+        reduced_cost, parent_matrix = matrix_reduction(graph)

```

Comment:
Does this do a first pass of matrix reduction?

It's a little nitpicky, but would like to explicitly follow the pseudocode steps:
```
1. Create matrix
2. Reduce matrix
3. Choose random starting node
4. Set Upper Bounds
5. BFS
```

Should generally look like this so that it's very obvious we're following the pseudocode:
```
1. Create Matrix
matrix = self.create_matrix(graph)

2. Reduce Matrix
reduced_cost, parent_matrix = matrix_reduction(matrix)

3. Choose random starting node
starting_node = random.choice( graph.keys() )
path.append( (starting_node[0], starting_node[2]) )

4. Set Upper Bounds
// Not really sure what this does

5. BFS
bfs(matrix)

```